### PR TITLE
Commands for working with external libraries

### PR DESCRIPTION
**Pull request by svor**
Adds several commands for working with external libraries:

- Get all external libraries (jars)
- Get children of library
- Get content of library's file

### CODE REVIEW

File: org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/externallibrary/LibraryChildrenCommand.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+import org.eclipse.lsp4j.jsonrpc.json.adapters.EitherTypeAdapterFactory;
+import org.eclipse.lsp4j.jsonrpc.json.adapters.EnumTypeAdapterFactory;
+
+/**
+ * A command to compute jar's children.
+ *
+ * @author Valeriy Svydenko
+ */
+public class LibraryChildrenCommand {
+  private static final Gson gson =
```
Comment (tsmaeder) [inline org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/externallibrary/LibraryChildrenCommand.java]: Gson instances should be thread safe, so we should stop making new instances and have a global one. I made an issue (https://github.com/eclipse/che-ls-jdt/issues/9) to be solved in a separate PR.

File: org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/externallibrary/LibraryNavigation.java [hunk 2]
```diff
@@ -0,0 +1,655 @@
...
+        JarEntryFile file = findJarFile((JarEntryDirectory) children, path);
+        if (file != null) {
+          return file;
+        }
+      }
+    }
+    return null;
+  }
+
+  private static IPackageFragmentRoot getPackageFragmentRoot(
```
Comment (tsmaeder) [inline org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/externallibrary/LibraryNavigation.java]: If the id is the handle identifier you can simply do: root= (IPackageFragmentRoot)JavaCore.create(id). No need to iterate over all roots.
In this way, we won't need the project to get the package fragment root for a library id. -> remove it from the parameters where applicable.

File: org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/externallibrary/LibraryNavigation.java [hunk 3]
```diff
@@ -0,0 +1,655 @@
...
+
+  /**
+   * Returns the hierarchical packages inside a given fragment or root.
+   *
+   * @param parent the parent package fragment root
+   * @param fragment the package to get the children for or 'null' to get the children of the root
+   * @param result Collection where the resulting elements are added
+   * @throws JavaModelException if fetching the children fails
+   */
+  private static void getHierarchicalPackageChildren(
```
Comment (tsmaeder) [inline org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/externallibrary/LibraryNavigation.java]: If you pass in a collection to be filled, it would be nicer to call the method "collect...."

File: org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/externallibrary/LibraryNavigation.java [hunk 4]
```diff
@@ -0,0 +1,655 @@
...
+      IPackageFragmentRoot parent,
+      IPackageFragment fragment,
+      Collection<Object> result,
+      IProgressMonitor pm)
+      throws JavaModelException {
+    IJavaElement[] children = parent.getChildren();
+    String prefix = fragment != null ? fragment.getElementName() + PERIOD : ""; // $NON-NLS-1$
+    int prefixLen = prefix.length();
+    for (IJavaElement aChildren : children) {
```
Comment (tsmaeder) [inline org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/externallibrary/LibraryNavigation.java]: singular of children is "child".

File: org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/externallibrary/LibraryNavigation.java [hunk 5]
```diff
@@ -0,0 +1,655 @@
...
+      IPackageFragmentRoot root, IProgressMonitor pm) throws JavaModelException {
+    ArrayList<Object> result = new ArrayList<>();
+    getHierarchicalPackageChildren(root, null, result, pm);
+    Object[] nonJavaResources = root.getNonJavaResources();
+    Collections.addAll(result, nonJavaResources);
+    return result.toArray();
+  }
+
+  /**
+   * Returns the hierarchical packages inside a given fragment or root.
```
Comment (tsmaeder) [inline org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/externallibrary/LibraryNavigation.java]: If I read this correctly, it this gives a tree of packages, with empty packages collapsed like so:

```
org.eclipse
  jdt.ls
  jdt.core
    IJavaElement.java
```
Since this is purely a display issue, I think in the long run we should move this concern to the client (browser) side.

File: org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/externallibrary/LibraryNavigation.java [hunk 6]
```diff
@@ -0,0 +1,655 @@
...
+        } catch (JavaModelException e) {
+          // ignore
+        }
+        return element;
+      }
+    }
+    return parent;
+  }
+
+  private static Comparator<JarEntry> comparator =
```
Comment (tsmaeder) [inline org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/externallibrary/LibraryNavigation.java]: This is a constant, right? Why not make it upper case and move it to the top of the file?

File: org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/externallibrary/LibraryNavigation.java (hunk 5 above)
Comment (svor) [reply org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/externallibrary/LibraryNavigation.java]: I think the description isn't so clear. Actually it is a tree where depth is 1 and collapsed default package (if it is a root of jar).In your example it will be:
``` 
org.eclipse
  jdt
```

File: org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/externallibrary/LibraryNavigation.java (hunk 4 above)
Comment (svor) [reply org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/externallibrary/LibraryNavigation.java]: yes, typo. Thx!

File: org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/externallibrary/LibraryNavigation.java (hunk 6 above)
Comment (svor) [reply org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/externallibrary/LibraryNavigation.java]: It's right. Fixed

File: org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/externallibrary/LibraryNavigation.java (hunk 3 above)
Comment (svor) [reply org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/externallibrary/LibraryNavigation.java]: fixed

File: org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/externallibrary/LibraryNavigation.java (hunk 2 above)
Comment (svor) [reply org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/externallibrary/LibraryNavigation.java]: Cool, thanks! Changed
